server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -215,8 +215,17 @@ public ClusterState execute(ClusterState currentState) {
                         final RepositoryMetadata updatedMetadata;
                         if (canUpdateInPlace(newRepositoryMetadata, existing)) {
                             if (repositoryMetadata.settings().equals(newRepositoryMetadata.settings())) {
-                                // Previous version is the same as this one no update is needed.
-                                return currentState;
+                                if (repositoryMetadata.generation() == RepositoryData.CORRUPTED_REPO_GEN) {
+                                    // If recreating a corrupted repository with the same settings, reset the corrupt flag (issue #89130).
+                                    // Setting the safe generation to unknown, so that a consistent generation is found.
+                                    repositoryMetadata = repositoryMetadata.withGeneration(
```

Comment:
I wonder if we should consider the removal of the corruption marker as a "reset" of the registered repository, like if it was deleted and then recreated. If it got corrupted it is possible that the repository content was modified under us, so we likely want to read the repository content again?

Reply:
@tlrx , thanks for the comment! For the moment, what I think is happening is that the repository gets re-read to find the safe generation to use. Which sounds like what would happen in the case of deletion + recreation, but I may be wrong? I wonder also if there would be any difference whether searchable snapshots are mounted or not in either case. Feel free to elaborate please.

Reply:
> For the moment, what I think is happening is that the repository gets re-read to find the safe generation to use. 

Agree. I was worried about resetting the generation to `UNKNOWN_REPO_GEN` without re-reading the repository UUID along with the pending generation, but if I understand correctly the repository UUID is also read and updated in that case too using `RepositoriesService.updateRepositoryUuidInMetadata()`.

I think we also want to prevent resetting the corruption marker if there are on-going snapshots or deletions? I expect them to quickly fail during finalization if the repository is corrupted but it is probably safer to ensure that when updating the cluster state to remove the corruption marker?

> I wonder also if there would be any difference whether searchable snapshots are mounted or not in either case.

Searchable snapshots use the repository UUID to locate a snapshot so as long as both stay the same it should work.

Reply:
Hi @tlrx , thanks for the review!

> I think we also want to prevent resetting the corruption marker if there are on-going snapshots or deletions? I expect them to quickly fail during finalization if the repository is corrupted but it is probably safer to ensure that when updating the cluster state to remove the corruption marker?

I see. Would adding a `ensureRepositoryNotInUse(currentState, request.name());` here (like in line 233 below) be enough? Or would you expect the check somewhere else?

Reply:
Yes, that's what I have in mind yes. We probably want a test for this too.

Reply:
Hi @tlrx , @original-brownbear , it took me some time to make a test with a concurrent snapshot but I finally did it and now resetting a corrupt repository fails if there are ongoing snapshots! It seems to me like a rare case but you never know indeed :) Feel free to review the new test, and unresolve this conversation if it is not sufficient!